File: packages/capacitor-notificare-scannables/android/src/main/java/re/notifica/scannables/capacitor/NotificareScannablesPlugin.kt
package re.notifica.scannables.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.scannables.NotificareScannables
import re.notifica.scannables.ktx.scannables
import re.notifica.scannables.models.NotificareScannable
import re.notifica.scannables.models.toJson

@CapacitorPlugin(name = "NotificareScannablesPlugin")
public class NotificareScannablesPlugin : Plugin(), NotificareScannables.ScannableSessionListener {
    override fun load() {
        logger.hasDebugLoggingEnabled = Notificare.options?.debugLoggingEnabled ?: false

        EventBroker.setup(this::notifyListeners)

        Notificare.scannables().removeListener(this)
        Notificare.scannables().addListener(this)
    }

    @PluginMethod
    public fun canStartNfcScannableSession(call: PluginCall) {
        call.resolve(JSObject().apply {
            put("result", Notificare.scannables().canStartNfcScannableSession)
        })
    }

    @PluginMethod
    public fun startScannableSession(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Cannot start a scannable session without an activity.")
            return
        }

        Notificare.scannables().startScannableSession(activity)
        call.resolve()
    }

    @PluginMethod
    public fun startNfcScannableSession(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Cannot start a scannable session without an activity.")
            return
        }

        Notificare.scannables().startNfcScannableSession(activity)
        call.resolve()
    }

    @PluginMethod
    public fun startQrCodeScannableSession(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Cannot start a scannable session without an activity.")
            return
        }

        Notificare.scannables().startQrCodeScannableSession(activity)
        call.resolve()
    }

    @PluginMethod
    public fun fetch(call: PluginCall) {
        val tag = call.getString("tag") ?: run {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Notificare.scannables().fetch(tag, object : NotificareCallback<NotificareScannable> {
            override fun onSuccess(result: NotificareScannable) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // region NotificareScannables.ScannableSessionListener

    override fun onScannableDetected(scannable: NotificareScannable) {
        try {
            EventBroker.dispatchEvent("scannable_detected", scannable.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the scannable_detected event.", e)
        }
    }

    override fun onScannableSessionError(error: Exception) {
        try {
            EventBroker.dispatchEvent("scannable_session_failed", JSObject().apply {
                put("error", error.localizedMessage)
            })
        } catch (e: Exception) {
            logger.error("Failed to emit the scannable_session_failed event.", e)
        }
    }

    // endregion
}

File: packages/capacitor-notificare-scannables/ios/Plugin/NotificareScannablesPlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareScannablesKit

@objc(NotificareScannablesPlugin)
public class NotificareScannablesPlugin: CAPPlugin {
    private var rootViewController: UIViewController? {
        get {
            UIApplication.shared.delegate?.window??.rootViewController
        }
    }
    
    public override func load() {
        addApplicationLaunchListener()

        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Notificare.shared.scannables().delegate = self
    }
    
    @objc func canStartNfcScannableSession(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.scannables().canStartNfcScannableSession
        ])
    }
    
    @objc func startScannableSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                call.reject("Cannot start a scannable session with a nil root view controller.")
                return
            }
            
            Notificare.shared.scannables().startScannableSession(controller: rootViewController)
            call.resolve()
        }
    }
    
    @objc func startNfcScannableSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Notificare.shared.scannables().startNfcScannableSession()
            call.resolve()
        }
    }
    
    @objc func startQrCodeScannableSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                call.reject("Cannot start a scannable session with a nil root view controller.")
                return
            }
            
            Notificare.shared.scannables().startQrCodeScannableSession(controller: rootViewController, modal: true)
            call.resolve()
        }
    }
    
    @objc func fetch(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("Missing 'tag' parameter.")
            return
        }
        
        Notificare.shared.scannables().fetch(tag: tag) { result in
            switch result {
            case let .success(scannable):
                do {
                    call.resolve([
                        "result": try scannable.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
}

extension NotificareScannablesPlugin: NotificareScannablesDelegate {
    public func notificare(_ notificareScannables: NotificareScannables, didDetectScannable scannable: NotificareScannable) {
        do {
            EventBroker.instance.dispatchEvent("scannable_detected", data: try scannable.toJson())
        } catch {
            logger.error("Failed to emit the scannable_detected event.", error: error)
        }
    }
    
    public func notificare(_ notificareScannables: NotificareScannables, didInvalidateScannerSession error: Error) {
        EventBroker.instance.dispatchEvent("scannable_session_failed", data: [
            "error": error.localizedDescription
        ])
    }
}

extension NotificareScannablesPlugin {
    private func addApplicationLaunchListener() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(didFinishLaunching),
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    private func removeApplicationLaunchListener() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    @objc private func didFinishLaunching() {
        removeApplicationLaunchListener()

        logger.hasDebugLoggingEnabled = Notificare.shared.options?.debugLoggingEnabled ?? false
    }
}
